contracts/interchain-token-service/src/contract/execute/mod.rs

```diff
@@ -1189,6 +1246,317 @@ mod tests {
+
+        let decimals = 18;
+        let supply = TokenSupply::Tracked(Uint256::one());
+        assert_err_contains!(
+            register_p2p_token_instance(
+                deps.as_mut(),
+                token_id(),
+                ChainNameRaw::try_from("bananas").unwrap(),
+                ethereum(),
+                decimals,
+                supply.clone()
+            ),
+            Error,
+            Error::ChainNotRegistered { .. }
+        );
+
+        assert_err_contains!(
+            register_p2p_token_instance(
+                deps.as_mut(),
+                token_id(),
+                ethereum(),
+                ChainNameRaw::try_from("bananas").unwrap(),
+                decimals,
+                supply.clone()
+            ),
+            Error,
+            Error::ChainNotRegistered { .. }
+        );
+    }
+
+    #[test]
+    fn register_p2p_token_should_init_balance_tracking() {
+        let mut deps = mock_dependencies();
+        init(&mut deps);
+
+        let decimals = 18;
+        assert_ok!(register_p2p_token_instance(
+            deps.as_mut(),
+            token_id(),
+            ethereum(),
+            ethereum(),
+            decimals,
+            TokenSupply::Untracked
+        ));
+
+        let supply = TokenSupply::Tracked(Uint256::one());
+        assert_ok!(register_p2p_token_instance(
+            deps.as_mut(),
+            token_id(),
+            solana(),
+            ethereum(),
+            decimals,
+            supply.clone(),
+        ));
+
+        let transfer_amount = Uint256::one().try_into().unwrap();
+
+        // initial supply is one token. first transfer should succeed
+        assert_ok!(do_transfer(
+            deps.as_mut(),
+            solana(),
+            ethereum(),
+            token_id(),
+            transfer_amount
+        ));
+
+        assert_err_contains!(
+            do_transfer(
+                deps.as_mut(),
+                solana(),
+                ethereum(),
+                token_id(),
+                transfer_amount
+            ),
+            Error,
+            Error::TokenSupplyInvariantViolated { .. }
+        );
+    }
+
+    // Below are various helper functions to assist with writing tests
+
+    fn its_address() -> nonempty::HexBinary {
+        HexBinary::from_hex(ITS_ADDRESS)
+            .unwrap()
+            .try_into()
+            .unwrap()
+    }
+
+    // most tests only need one token id
+    fn token_id() -> TokenId {
+        TokenId::new([7u8; 32])
+    }
+
+    fn solana() -> ChainNameRaw {
+        SOLANA.try_into().unwrap()
+    }
+
+    fn xrpl() -> ChainNameRaw {
+        XRPL.try_into().unwrap()
+    }
+
+    fn ethereum() -> ChainNameRaw {
+        ETHEREUM.try_into().unwrap()
+    }
+
+    fn cc_id(source_chain: ChainNameRaw) -> CrossChainId {
+        CrossChainId {
+            source_chain,
+            message_id: HexTxHashAndEventIndex::new([1u8; 32], 0u32).into(),
+        }
+    }
+
+    fn do_transfer(
```

Comment:
```suggestion
    fn transfer_token(
```
`do_*` naming feels inconsistent with other functions